### PR TITLE
Feat: Support for dimension during export Journal Entry and purchase Invoice

### PR DESCRIPTION
**Pull request by Ashutosh619-sudo**
…Invoice

### Description
Please add PR description here, add screenshots if needed

## Clickup
Please add link here
https://app.clickup.com/1864988/v/l/6-901604887942-1

## Release Notes

- **New Features**
	- Introduced dimension handling for journal entries and purchase invoices, allowing for the storage of Business Central dimensions and related error logs.
	- Added functionality to post dimension line data for purchase invoices and journal entries.
	- Enhanced methods to construct and manage dimension payloads for improved integration with Business Central.

- **Bug Fixes**
	- Enhanced error management for dimension processing, ensuring logs are maintained for any issues encountered.

- **Documentation**
	- Updated help texts for new fields related to dimensions and error logs in journal entry and purchase invoice models.

### CODE REVIEW

File: apps/business_central/exports/base_model.py [hunk 1]
```diff
@@ -182,3 +182,43 @@ def get_location_id(accounting_export: AccountingExport, lineitem: Expense):
             if mapping:
                 location_id = mapping.destination.destination_id
         return location_id
+
+
+def get_dimension_object(accounting_export: AccountingExport, lineitem: Expense):
```
Comment (ruuushhh) [inline apps/business_central/exports/base_model.py:187]: move this inside BaseModel

File: apps/business_central/exports/purchase_invoice/tasks.py [hunk 2]
```diff
@@ -61,21 +63,24 @@ def __construct_purchase_invoice(self, body: PurchaseInvoice, lineitems: List[Pu
...
+        return purchase_invoice_payload, batch_purchase_invoice_lineitem_payload, dimension_set_line_payloads
 
     def post(self, accounting_export, item, lineitem):
         '''
         Export the Journal Entry to Business Central.
         '''
 
-        purchase_invoice_payload, batch_purchase_invoice_payload = self.__construct_purchase_invoice(item, lineitem)
+        purchase_invoice_payload, batch_purchase_invoice_payload, dimension_set_line_payloads = self.__construct_purchase_invoice(item, lineitem)
         logger.info('WORKSPACE_ID: {0}, ACCOUNTING_EXPORT_ID: {1}, PURCHASE_INVOICE_PAYLOAD: {2}, BATCH_PURCHASE_INVOICE_PAYLOAD: {3}'.format(accounting_export.workspace_id, accounting_export.id, purchase_invoice_payload, batch_purchase_invoice_payload))
```
Comment (ruuushhh) [inline apps/business_central/exports/purchase_invoice/tasks.py:74]: add dimensions payload to logger as well

File: apps/business_central/exports/purchase_invoice/tasks.py [hunk 3]
```diff
@@ -61,21 +63,24 @@ def __construct_purchase_invoice(self, body: PurchaseInvoice, lineitems: List[Pu
...
         business_central_credentials = BusinessCentralCredentials.get_active_business_central_credentials(accounting_export.workspace_id)
         # Establish a connection to Business Central
         business_central_connection = BusinessCentralConnector(business_central_credentials, accounting_export.workspace_id)
 
         response = business_central_connection.post_purchase_invoice(purchase_invoice_payload, batch_purchase_invoice_payload)
 
+        if dimension_set_line_payloads:
+            dimension_line_responses = business_central_connection.post_dimension_lines(dimension_set_line_payloads, response['bulk_post_response']['response']['body']['id'],'PURCHASE_INVOICE')
```
Comment (ruuushhh) [inline apps/business_central/exports/purchase_invoice/tasks.py:82]: what happens when first post call fails, do we do the dimension post call still?

Comment (ruuushhh) [inline apps/business_central/exports/purchase_invoice/tasks.py:82]: let's add this response to purchase invoice  response

File: apps/business_central/exports/base_model.py (hunk 1 above)
Comment (Ashutosh619-sudo) [reply apps/business_central/exports/base_model.py:187]: done

File: apps/business_central/exports/purchase_invoice/tasks.py (hunk 2 above)
Comment (Ashutosh619-sudo) [reply apps/business_central/exports/purchase_invoice/tasks.py:74]: done

File: apps/business_central/exports/purchase_invoice/tasks.py (hunk 3 above)
Comment (Ashutosh619-sudo) [reply apps/business_central/exports/purchase_invoice/tasks.py:82]: done

File: apps/business_central/utils.py [hunk 4]
```diff
@@ -291,6 +295,32 @@ def post_purchase_invoice(self, purchase_invoice_payload, purchase_invoice_linei
...
+        exception_response = []
+
+        for dimension_line_payload in dimension_line_payloads:
+            exported_module_id = dimension_line_payload['exported_module_id']
+            dimension_line_payload.pop('exported_module_id')
+            try:
+                if export_module_type == 'JOURNAL_ENTRY':
+                    response = self.connection.journal_line_items.post_journal_entry_dimensions(journal_line_item_id=dimension_line_payload['parentId'], data=dimension_line_payload)
+                else:
+                    response = self.connection.purchase_invoice_line_items.post_purchase_invoice_dimensions(purchase_invoice_item_id=dimension_line_payload['parentId'], data=dimension_line_payload)
```
Comment (ruuushhh) [inline apps/business_central/utils.py:311]: add this to accounting export detail with new key

File: apps/business_central/exports/journal_entry/models.py [hunk 5]
```diff
@@ -83,6 +85,8 @@ class JournalEntryLineItems(BaseExportModel):
     invoice_date = CustomDateTimeField(help_text='date of invoice')
     document_number = TextNotNullField(help_text='document number of the invoice')
     journal_entry = models.ForeignKey(JournalEntry, on_delete=models.PROTECT, help_text='Journal Entry reference', related_name='journal_entry_lineitems')
+    dimensions = JSONField(null=True, help_text='Business Central dimensions')
```
Comment (ashwin1111) [inline apps/business_central/exports/journal_entry/models.py:88]: default=list and remove null

File: apps/business_central/exports/journal_entry/tasks.py [hunk 6]
```diff
@@ -99,6 +116,35 @@ def post(self, accounting_export, item, lineitem):
...
+    
+    def construct_dimension_set_line_payload(self, dimensions: list, exported_response: dict):
+        """
+        construct payload for setting dimension for Purchase Invoice
+        """
+
+        dimension = dimensions[0]
+
+        #we are only supporting grouping by expense for JE, so only two payloads will always exist
+        # one for each line
```
Comment (ashwin1111) [inline apps/business_central/exports/journal_entry/tasks.py:128]: not true, dimensions can be n for 1 expense

File: apps/business_central/utils.py [hunk 7]
```diff
@@ -291,6 +295,32 @@ def post_purchase_invoice(self, purchase_invoice_payload, purchase_invoice_linei
...
+            except Exception as exception:
+                if export_module_type == 'JOURNAL_ENTRY':
+                    lineitem = JournalEntryLineItems.objects.get(id=exported_module_id)
```
Comment (ashwin1111) [inline apps/business_central/utils.py:314]: use parent id as well in query

File: apps/business_central/exports/purchase_invoice/tasks.py [hunk 8]
```diff
@@ -49,33 +50,49 @@ def __construct_purchase_invoice(self, body: PurchaseInvoice, lineitems: List[Pu
...
-        purchase_invoice_payload, batch_purchase_invoice_payload = self.__construct_purchase_invoice(item, lineitem)
+        purchase_invoice_payload, batch_purchase_invoice_payload, dimensions = self.__construct_purchase_invoice(item, lineitem)
         logger.info('WORKSPACE_ID: {0}, ACCOUNTING_EXPORT_ID: {1}, PURCHASE_INVOICE_PAYLOAD: {2}, BATCH_PURCHASE_INVOICE_PAYLOAD: {3}'.format(accounting_export.workspace_id, accounting_export.id, purchase_invoice_payload, batch_purchase_invoice_payload))
         business_central_credentials = BusinessCentralCredentials.get_active_business_central_credentials(accounting_export.workspace_id)
         # Establish a connection to Business Central
         business_central_connection = BusinessCentralConnector(business_central_credentials, accounting_export.workspace_id)
 
         response = business_central_connection.post_purchase_invoice(purchase_invoice_payload, batch_purchase_invoice_payload)
 
+        try:
```
Comment (ashwin1111) [inline apps/business_central/exports/purchase_invoice/tasks.py:83]: remove

File: apps/business_central/utils.py [hunk 9]
```diff
@@ -291,6 +295,32 @@ def post_purchase_invoice(self, purchase_invoice_payload, purchase_invoice_linei
...
+                else:
+                    lineitem = PurchaseInvoiceLineitems.objects.get(id=exported_module_id)
+
+                lineitem.dimension_error_log = str(exception.response)
+                lineitem.save()
+                exception_response.append(str(exception.response))
+        
+        return exception_response
```
Comment (ashwin1111) [inline apps/business_central/utils.py:322]: not required

File: apps/business_central/utils.py [hunk 10]
```diff
@@ -291,6 +295,32 @@ def post_purchase_invoice(self, purchase_invoice_payload, purchase_invoice_linei
...
+    def post_dimension_lines(self, dimension_line_payloads: List[Dict], export_module_type: str):
+        """
+        Post dimension line for purchase invoice line and Journal line item
+        """
+        exception_response = []
+
+        for dimension_line_payload in dimension_line_payloads:
+            exported_module_id = dimension_line_payload['exported_module_id']
+            dimension_line_payload.pop('exported_module_id')
+            try:
```
Comment (ashwin1111) [inline apps/business_central/utils.py:307]: return successful response

Comment (ashwin1111) [reply apps/business_central/utils.py:307]: append all success response and return

File: apps/business_central/utils.py [hunk 11]
```diff
@@ -291,6 +295,32 @@ def post_purchase_invoice(self, purchase_invoice_payload, purchase_invoice_linei
...
+                if export_module_type == 'JOURNAL_ENTRY':
+                    response = self.connection.journal_line_items.post_journal_entry_dimensions(journal_line_item_id=dimension_line_payload['parentId'], data=dimension_line_payload)
+                else:
+                    response = self.connection.purchase_invoice_line_items.post_purchase_invoice_dimensions(purchase_invoice_item_id=dimension_line_payload['parentId'], data=dimension_line_payload)
+            except Exception as exception:
+                if export_module_type == 'JOURNAL_ENTRY':
```
Comment (ashwin1111) [inline apps/business_central/utils.py:313]: some success and some failed to be handled

File: apps/business_central/exports/purchase_invoice/tasks.py [hunk 12]
```diff
@@ -49,33 +50,49 @@ def __construct_purchase_invoice(self, body: PurchaseInvoice, lineitems: List[Pu
...
+            if dimensions:
+                dimension_set_line_payloads = self.construct_dimension_set_line_payload(dimensions, response['bulk_post_response']['responses'])
+                logger.info('WORKSPACE_ID: {0}, ACCOUNTING_EXPORT_ID: {1}, DIMENSION_SET_LINE_PAYLOADS: {2}'.format(accounting_export.workspace_id, accounting_export.id, dimension_set_line_payloads))
+                dimension_line_responses = business_central_connection.post_dimension_lines(
+                    dimension_set_line_payloads, 'PURCHASE_INVOICE'
+                )
+                response['dimension_line_responses'] = dimension_line_responses
+        except Exception as exception:
+            lineitem.dimension_error_log = str(exception.response)
```
Comment (ashwin1111) [inline apps/business_central/exports/purchase_invoice/tasks.py:92]: store success response in line table, not needed in accounting export

File: apps/business_central/exports/journal_entry/tasks.py (hunk 6 above)
Comment (Ashutosh619-sudo) [reply apps/business_central/exports/journal_entry/tasks.py:128]: done

File: apps/business_central/exports/purchase_invoice/tasks.py (hunk 8 above)
Comment (Ashutosh619-sudo) [reply apps/business_central/exports/purchase_invoice/tasks.py:83]: done

File: apps/business_central/exports/purchase_invoice/tasks.py (hunk 12 above)
Comment (Ashutosh619-sudo) [reply apps/business_central/exports/purchase_invoice/tasks.py:92]: done

File: apps/business_central/utils.py [hunk 13]
```diff
@@ -291,6 +295,50 @@ def post_purchase_invoice(self, purchase_invoice_payload, purchase_invoice_linei
...
+                lineitem.dimension_success_log = str(response)
+                lineitem.save()
+
+            except Exception as exception:
+                lineitem = get_lineitem_by_id(export_module_type, exported_module_id)
+                error_message = str(getattr(exception, 'response', exception))
+                lineitem.dimension_error_log = error_message
+                lineitem.save()
+
+        return exception_response
```
Comment (ruuushhh) [inline apps/business_central/utils.py:339]: no need to return anything

**Review by ruuushhh — APPROVED** [review]

File: apps/business_central/utils.py [hunk 14]
```diff
@@ -291,6 +295,50 @@ def post_purchase_invoice(self, purchase_invoice_payload, purchase_invoice_linei
...
+                        journal_line_item_id=dimension_line_payload['parentId'],
+                        data=dimension_line_payload
+                    )
+                else:
+                    response = self.connection.purchase_invoice_line_items.post_purchase_invoice_dimensions(
+                        purchase_invoice_item_id=dimension_line_payload['parentId'],
+                        data=dimension_line_payload
+                    )
+
+                lineitem = get_lineitem_by_id(export_module_type, exported_module_id)
```
Comment (ruuushhh) [inline apps/business_central/utils.py:329]: move this out of try catch repeated code

File: apps/business_central/utils.py [hunk 15]
```diff
@@ -291,6 +295,50 @@ def post_purchase_invoice(self, purchase_invoice_payload, purchase_invoice_linei
...
+                lineitem.dimension_success_log = str(response)
+                lineitem.save()
+
+            except Exception as exception:
+                lineitem = get_lineitem_by_id(export_module_type, exported_module_id)
```
Comment (ruuushhh) [inline apps/business_central/utils.py:334]: same here

**Review by ruuushhh — APPROVED** [review]

**Pull request merged**